ray_maple_workflow.py

```diff
@@ -107,26 +117,48 @@ def add_image_name(row: Dict[str, Any]) -> Dict[str, Any]:
-                fn=ray_tile_and_stitch_util.tile_image,
-                fn_kwargs={"config": config},
-                concurrency=args.concurrency).drop_columns(["mask"]).map(
-                    fn=ray_infer_tiles.MaskRCNNPredictor,
-                    fn_constructor_kwargs={"config": config},
-                    concurrency=args.concurrency).groupby("image_name").map_groups(
-                        ray_tile_and_stitch_util.stitch_shapefile,
-                        concurrency=args.concurrency).map(
-                            fn=ray_write_shapefiles.WriteShapefiles,
-                            fn_constructor_kwargs={"config": config},
-                            concurrency=args.concurrency)
-
-    print("Ray pipeline finished: ", ray_dataset.schema())
+    concurrency = args.concurrency
+
+    print("Starting MAPLE Ray pipeline...")
+    print("""This pipeline will:
+          - load geotiffs into ray dataset
+          - calculate a watermask for each image
+          - tile each watermasked image
+          - perform inference on each tile
+          - stitch the tile inference results to get the inference results for each image
+          - write the inference results for each image to shapefiles 
+          """)
+
+    # 0. Load geotiffs into ray dataset
+    dataset = create_geotiff_images_dataset(config).map(add_image_name, concurrency=concurrency)
+
+    # 1. Start calculating watermask
+    dataset_with_water_mask = dataset.map(fn=ray_image_preprocessing.cal_water_mask,
+                                          fn_kwargs={"config": config}, concurrency=args.concurrency)
+
+    # 2. Start tiling image
+    image_tiles_dataset = dataset_with_water_mask.flat_map(
+        fn=ray_tile_and_stitch_util.tile_image, fn_kwargs={"config": config}, concurrency=args.concurrency)
+    image_tiles_dataset = image_tiles_dataset.drop_columns(["mask"])
+
+    # 3. Start inferencing
+    inferenced_dataset = image_tiles_dataset.map(
+        fn=ray_infer_tiles.MaskRCNNPredictor, fn_constructor_kwargs={"config": config}, concurrency=concurrency)
+
+    # 4. Start stitching
+    data_per_image = inferenced_dataset.groupby(
+        "image_name").map_groups(ray_tile_and_stitch_util.stitch_shapefile, concurrency=args.concurrency)
+
+    # 5. Write shapefiles
+    # Create the output directory if it doesn't exist.
+    # TODO - do something similar for GCP directories, we have the code to create local
+    # dirs if they don't exist (ex. mpl_workflow_create_dir_struct.py), it'd be great to
+    # add support for creating GCP directories if they don't exist.
+    if config.GCP_FILESYSTEM is None:
+            create_directory_if_not_exists(config.RAY_OUTPUT_SHAPEFILES_DIR)
+    shapefiles_dataset = data_per_image.map(
+        fn=ray_write_shapefiles.WriteShapefiles, fn_constructor_kwargs={"config": config}, concurrency=concurrency)
+    print("MAPLE Ray pipeline finished, done writing shapefiles", shapefiles_dataset.schema())
```

Comment:
You'll actually want to call materialize here instead of schema. Materialize actually runs the pipeline for all the rows in the dataset, while my understanding is that schema only materializes a single row in order to get the schema of the data set.

Reply:
How'd you find out the schema only materializes a single row? I updated the code, I'm just curious

Reply:
When I was prototyping with running on more than 1 image, if you did .schema() it would only produce results for a single row